subsys/usb/device_next/class/usbd_hid.c

```diff
@@ -0,0 +1,743 @@
+	/* At least report subordinate descriptor is required. */
+	struct subordinate_info sub[HID_SUBORDINATE_DESC_NUM];
+} __packed;
+
+struct usbd_hid_descriptor {
+	struct usb_if_descriptor if0;
+	struct hid_descriptor hid;
+	struct usb_ep_descriptor in_ep;
+	struct usb_ep_descriptor hs_in_ep;
+	struct usb_ep_descriptor out_ep;
+	struct usb_ep_descriptor hs_out_ep;
+
+	struct usb_if_descriptor if0_1;
+	struct usb_ep_descriptor alt_hs_in_ep;
+	struct usb_ep_descriptor alt_hs_out_ep;
+};
+
+enum {
+	HID_DEV_CLASS_ENABLED,
+};
+
+struct hid_device_data {
+	struct usbd_hid_descriptor *const desc;
+	struct usbd_class_data *c_data;
+	struct net_buf_pool *const pool_out;
+	struct net_buf_pool *const pool_in;
+	const struct hid_device_ops *ops;
+	const uint8_t *rdesc;
+	size_t rsize;
+	const struct usb_desc_header **const fs_desc;
+	const struct usb_desc_header **const hs_desc;
+	atomic_t state;
+	struct k_sem in_sem;
+	struct k_work output_work;
+	uint8_t idle_rate;
+	uint8_t protocol;
+};
+
+static uint8_t hid_get_in_ep(struct usbd_class_data *const c_data)
+{
+	const struct device *dev = usbd_class_get_private(c_data);
+	struct hid_device_data *ddata = dev->data;
+	struct usbd_hid_descriptor *desc = ddata->desc;
+
+	return desc->in_ep.bEndpointAddress;
+}
+
+static inline uint8_t hid_get_out_ep(struct usbd_class_data *const c_data)
+{
+	const struct device *dev = usbd_class_get_private(c_data);
+	struct hid_device_data *ddata = dev->data;
+	struct usbd_hid_descriptor *desc = ddata->desc;
+
+	return desc->out_ep.bEndpointAddress;
+}
+
+static int usbd_hid_request(struct usbd_class_data *const c_data,
+			    struct net_buf *const buf, const int err)
+{
+	struct usbd_contex *uds_ctx = usbd_class_get_ctx(c_data);
+	const struct device *dev = usbd_class_get_private(c_data);
+	struct hid_device_data *ddata = dev->data;
+	const struct hid_device_ops *ops = ddata->ops;
+	struct udc_buf_info *bi;
+
+	bi = udc_get_buf_info(buf);
+
+	if (bi->ep == hid_get_out_ep(c_data)) {
+		if (ops->output_report != NULL) {
+			if (err == 0) {
+				ops->output_report(dev, buf->len, buf->data);
+			}
+
+			k_work_submit(&ddata->output_work);
+		}
+	}
+
+	if (bi->ep == hid_get_in_ep(c_data)) {
+		if (ops->input_report_done != NULL) {
+			ops->input_report_done(dev);
+		} else {
+			k_sem_give(&ddata->in_sem);
+		}
+	}
+
+	return usbd_ep_buf_free(uds_ctx, buf);
+}
+
+static int handle_set_idle(const struct device *dev,
+			   const struct usb_setup_packet *const setup)
+{
+	const uint32_t duration = HID_GET_IDLE_DURATION(setup->wValue);
+	const uint8_t id = HID_GET_IDLE_ID(setup->wValue);
+	struct hid_device_data *const ddata = dev->data;
+	const struct hid_device_ops *ops = ddata->ops;
+
+	if (id == 0U) {
+		/* Only the common idle rate is stored. */
+		ddata->idle_rate = duration;
+	}
+
+	LOG_DBG("Set Idle, Report ID %u Duration %u", id, duration);
+
+	if (ops->set_idle != NULL) {
+		ops->set_idle(dev, id, duration * 4UL);
+	} else {
+		errno = -ENOTSUP;
```

Comment:
We shouldn't end up with this error if `id == 0U`

Reply:
Why not? Id 0 applies to all reports and Set Report is not always required, so we need to issue protocol error if there is no callback.

Reply:
Then shouldn't we also return error for `id == 0` in `handle_get_idle` if there is no callback? (for consistency).
Also, currently we update our common idle rate even though we return protocol error here.

btw. This is `set_idle`

Reply:
Then I have to drop legacy API support.

Reply:
I think it's worth to keep the legacy API support (even if it has some additional limitations)